lib/compat.php

```diff
@@ -172,6 +172,98 @@ function gutenberg_override_reusable_block_post_type_labels() {
 }
 add_filter( 'post_type_labels_wp_block', 'gutenberg_override_reusable_block_post_type_labels', 10, 0 );
 
+/**
+ * Registers the user meta property responsible for storing preferences for the
+ * data script persistence middleware. The meta is explicitly registered so that
+ * it is made accessible for read and write via the REST API.
+ */
+function gutenberg_register_data_persistence_user_meta() {
+	global $wpdb;
+	register_meta(
+		'user',
+		$wpdb->get_blog_prefix() . 'data_persistence',
+		array(
+			'type'         => 'object',
+			'single'       => true,
+			'show_in_rest' => array(
+				'name'   => 'data_persistence',
+				'type'   => 'object',
+				'schema' => array(
+					'type'                 => 'object',
+					'properties'           => array(),
+					'additionalProperties' => true,
+				),
+			),
+		)
+	);
+}
+add_action( 'init', 'gutenberg_register_data_persistence_user_meta' );
+
+/**
+ * Enqueues inline scripts used to provide a custom storage interface for the
+ * data script persistence middleware. The custom storage implementation stores
+ * persisted values using user meta, bootstrapped for initial read at load-time,
+ * and updated via the REST API on change.
+ */
+function gutenberg_user_settings_data_persistence_inline_script() {
+	global $wp_scripts, $wpdb;
+
+	$user_id = get_current_user_id();
+	if ( empty( $user_id ) ) {
+		/*
+		 * In most cases, it should be expected that this would only be reached
+		 * while logged in and viewing the block editor. However, there may be
+		 * edge cases where a plugin is calling to enqueue block assets outside
+		 * the editor and when a user may not be logged in.
+		 */
+		return;
+	}
+
+	$persisted_value = get_user_meta( $user_id, $wpdb->get_blog_prefix() . 'data_persistence', true );
+	if ( empty( $persisted_value ) ) {
+		/*
+		 * If there's no explicit metadata assigned, fall back to a value which
+		 * was persisted using browser storage, prior to user meta persistence.
+		 */
+		$persisted_value = sprintf( 'localStorage.getItem( "WP_DATA_USER_%s" );', $user_id );
+	} else {
+		/*
+		 * Otherwise, encode the string value for interpolation in the storage
+		 * implementation script. The first `json_encode` will is responsible
+		 * for producing a JSON encoding of the persisted meta object, and the
+		 * second will apply quoting to that string result.
+		 */
+		$persisted_value = wp_json_encode( wp_json_encode( $persisted_value ) );
+	}
+
+	$persistence_script = <<<JS
+( function() {
+	wp.data.use( wp.data.plugins.persistence, {
+		storage: {
+			getItem: function() {
+				return {$persisted_value};
+			},
+			setItem: function( key, value ) {
```

Comment:
This doesn't seem to impact the return value of `getItem`, does it mean that we only use `getItem` to get the "initial value"?

Reply:
Yeah... I feel we should probably use `wp.apiFetch` to get the value, and if that fails then fallback to localStorage. Which means that then we should be able to migrate this whole logic to JS instead of PHP, most likely in the [data/src/plugins/persistence](https://github.com/WordPress/gutenberg/tree/trunk/packages/data/src/plugins/persistence) files directly :thinking:

Reply:
We can't update the default persistence plugin because the data module is a generic package with no backend/REST API dependency. 

I'm personally fine with the current implementation if the contract is that the `getItem` is about the initial value. If it's not we can just add a local variable we update when calling. `setItem` and return in in `getItem` (in addition to the API request in setItem)

Reply:
Added a local var :+1: